rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+With the above new implementation, we should be able to continue to use `StorageMap` as before:
+
+```rust
+storage {
+    my_map: StorageMap<u64, u64> = StorageMap {}
+}
+
+storage.my_map.insert(0, 0);
+let x = storage.my_map.get(0);
+storage.my_map.remove(0);
+```
+
+A similar approach can be followed to re-implement `StorageVec` and other dynamic storage types.
+
+## Dynamic Storage Types in Structs
+
+The code below will now be valid after implementing all of the above:
+
+```rust
+MyStruct {
+    map1: StorageMap<u64, u64>,
+    map2: StorageMap<u64, u64>,
+}
+
+storage {
+    s: Mystruct = MyStruct {
+        map1: StorageMap { },
+        map2: StorageMap { },
+    }
+}
+
+storage.s.map1.insert(0, 0);
+let x = storage.s.map2.get(0);
+```
+
+It will also be possible to pass a `StorageMap` to a function by passing a `StorageHandler` to it as follows:
+
+```rust
+storage {
+    my_map: StorageMap<u64, u64> = StorageMap {}
+}
+
+fn foo(map: StorageHandler<StorageMap<u64, u64>>) {
+    map.insert(0, 0);
+}
+
+fn bar() {
+    foo(storage.my_map); 
+}
+```
+
+Sway code using storage maps and vectors will not look any different than before after this RFC is implemented.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+There are two main drawbacks for the approach proposed above:
+
+1. Reading and writing a storage variable is now a bit more complicated and less ergonomic than before because we now have to call `store` and `get` manually. We also have to handle the `Option` returned by `get` which can be quite verbose. We should consider making this more ergonomic by introducing a simpler syntax that de-sugars to the API calls above. For example, we could de-sugar something like `storage.x = 42` to `storage.x.store(42)`. We could also consider making `get` return the data directly instead of an `Option` if we completely disallow storage slots corresponding to storage variables from being invalidated (For example, remove `std::storage::clear`).
+
+2. The second drawback is related to the fact that structs take up more storage slots than they actually need. This behavior is **not** new. There is a tradeoff here to be considered. If we make structs tightly packed, then manipulating a single field becomes more expensive (potential read-modify-write patterns or fields "spilling" over to the next slot), but the number of storage slots required may become lower. We might want to consider a `#[packed]` attribute in the future to let the user decide on one way v.s. the other. Another thing to consider with tightly-packed structs is that the type `StorageHandler` would then also require an additional field that points to a particular location _inside_ a given slot (i.e. which byte to start reading from or writing to).
```

Comment:
> Another thing to consider with tightly-packed structs is that the type `StorageHandler` would then also require an additional field that points to a particular location _inside_ a given slot (i.e. which byte to start reading from or writing to).

I can see how there would need to be different behaviour for `get` and `store` based on whether a struct is packed or  unpacked, however these implementation differences could be resolved at compile-time no?

I guess that would require representing the offset with a zero-sized type (e.g. `StorageHandler<Offset<Bytes3, Foo>>` or similar) which would break a lot of the usability as you could no longer pass the handle to functions that take `StorageHandler<Foo>` :thinking: Maybe that's fine though, as I'd imagine packed types would tend to just store primitive data anyway, so there may be less of a need to pass around `StorageHanlder`s for those fields anyway?

Reply:
Packed is now the default. Future discussion can consider unpacking sturcts via an annotation.